resources/build/build-utils.sh

```diff
@@ -288,87 +288,335 @@ fi
+  shift
+  # describe each target, action, and option possibility
+  while [[ $# -gt 0 ]]; do
+    local key="$1"
+    local value="$(echo "$key" | cut -d" " -f 1 -)"
+    local description=
+    if [[ $key =~ [[:space:]] ]]; then
+      description=$(_builder_trim "$(echo "$key" | cut -d" " -f 2- -)")
+    fi
+    if [[ $value =~ ^: ]]; then
+      _builder_targets+=($value)
+    elif [[ $value =~ ^-- ]]; then
+      # Look for a shorthand version of the option
+      if [[ $value =~ , ]]; then
+        local option_long="$(echo "$value" | cut -d, -f 1 -)"
+        local option_short="$(echo "$value" | cut -d, -f 2 -)"
+        _builder_options+=($option_long)
+        _builder_options_short[$option_short]="$option_long"
+        value="$option_long, $option_short"
+      else
+        _builder_options+=($value)
+      fi
+    else
+      _builder_actions+=($value)
+    fi
+
+    if [[ -z "${description}" ]]; then
+      description=$(_builder_get_default_description "$value")
+    fi
+    _builder_params[${value}]="$description"
+
+    shift
+  done
+
+  # We'll always add a :project if no target is specified
+  if (( ! ${#_builder_targets[@]} )); then
+    _builder_targets+=(:project)
+    _builder_params[\:project]=$(_builder_get_default_description ":project")
+  fi
+}
+
+_builder_get_default_description() {
+  local description=
+  local value="$1"
+  # default descriptions for common build actions, targets, and options
+  case "$value" in
+    clean)     description="remove build/ folder and build artifacts" ;;
+    configure) description="install dependencies, e.g. npm" ;;
+    build)     description="build target(s)" ;;
+    test)      description="run automated tests" ;;
+    ":project")  description="this project" ;;
+    :app)      description="main app" ;;
+    :engine)   description="engine module" ;;
+    :module)   description="this module" ;;
+    :tools)    description="build tools for this project" ;;
+    --debug)   description="debug build" ;;
+  esac
+  echo "$description"
+}
+
 # Initializes a build.sh script, parses command line. Will abort the script if
-# invalid parameters are passed in.
+# invalid parameters are passed in. This is a shorthand for builder_describe
+# followed by builder_parse, useful for simple scripts which don't need
+# additional detail in parameters or multiple targets.
 #
 # Usage:
-#   builder_init "action1 action2 ..." "$@"
+#   builder_init "action1 action2" "$@"
 # Parameters
-#   1: actions    space-separated list of possible actions
-#   [2: targets]  space-separated list of possible targets [TODO]
-#   3: $@         command-line arguments
+#   1: _builder_actions    space-separated list of possible actions
+#   2: $@         command-line arguments
 builder_init() {
-  local possible_actions=($1)
+  _builder_actions=($1)
+  _builder_targets=(:project)
+  _builder_options=()
+  declare -A -g _builder_params
+  declare -A -g _builder_options_short
   shift
+  builder_parse "$@"
+}
```

Comment:
Uh, the comment says "shorthand for builder_describe", but I don't see it called anywhere within this shell function.  Am I missing something?

Reply:
It effectively does the same thing as `builder_describe` but no, it currently doesn't call it. Should we eliminate this shortcut and always use `builder_describe`? It's not like it's a huge burden to fixup existing references at this stage. This would simplify future maintenance...

Reply:
> It effectively does the same thing as builder_describe but no, it currently doesn't call it.

OK, my ability to parse shell script may be a bit limited for the structures seen here, so feel free to correct me if I'm wrong on my interpretations here.

As best as I can tell... that statement only holds true for limited scenarios.  `builder_describe` inspects each 'key' of the description, parsing it to determine whether or not it's an action, an option, or a target.  As best as I can tell... `builder_init` says "these are all actions and nothing else" - I see nothing that double checks for `--option` or `:target` form specifications.  It just automatically says "there are no options and there's only one target - `:project`.  There's no explicit note of these caveats within the function's documentation comment, though.

Am I wrong, with `builder_init` able to take in something like this instead for the example?

```bash
builder_init "clean build test :app :tools --power --zoom"
```

Or will this _only_ work with `builder_describe`?  (Obviously, descriptions for the custom names may be missing, but let's ignore that part.)  The documentation currently present, as written, implies that `builder_init` should be able to handle such an initialization string, but I'm not confident of this after inspection of the code.

Reply:
Yeah, it's limited to defining actions, as it was the old style. I'm going to go ahead and tidy this up by only having a single approach -- there's no point in having both!

I may end up renaming `build_describe` to `builder_init` once I have done that -- it's a clearer name.

Reply:
OK, changed, but didn't rename `builder_describe` -- it's good as is.